File: examples/bases-prototype/observe.py
"""Capture real Obsidian Reading-view DOM; requires enabled Obsidian CLI.

Only use disposable vaults built by prototype.py. This is an evaluation harness,
not a plugin, and uses private UI APIs that may change between Obsidian releases.
"""
import argparse
import json
from pathlib import Path
import subprocess

CLI = "/Applications/Obsidian.app/Contents/MacOS/obsidian-cli"


class Observer:
    def __init__(self, vault, cli=CLI):
        self.vault = vault
        self.cli = cli

    def command(self, *args):
        # This CLI version requires the vault selector before the command.
        result = subprocess.run([self.cli, "vault=" + self.vault, *args], text=True, capture_output=True, timeout=60)
        if result.returncode or result.stdout.startswith("Error:") or "not enabled" in result.stdout:
            raise RuntimeError(result.stdout + result.stderr)
        return result.stdout.strip()

    def js(self, code):
        guarded = "(()=>{if(app.vault.getName()!==" + json.dumps(self.vault) + ")throw Error('CLI targeted the wrong vault');return (" + code + ");})()"
        output = self.command("eval", "code=" + guarded)
        if not output.startswith("=> "):
            raise RuntimeError(output)
        return json.loads(output[3:])

    def open(self, path):
        return self.js("""(async()=>{
          const start=performance.now();
          const w=require('@electron/remote').getCurrentWindow(); w.show(); w.focus();
          await app.workspace.activeLeaf.setViewState({type:'markdown',state:{file:PATH,mode:'preview'}});
          await new Promise(r=>setTimeout(r,300));
          return JSON.stringify({path:app.workspace.getActiveFile().path,openWith300msWait:performance.now()-start});
        })()""".replace("PATH", json.dumps(path)))

    def capture(self, embed):
        return self.js(r"""(async()=>{
          const selector='.internal-embed[src='+JSON.stringify(EMBED)+']';
          let el=document.querySelector(selector);
          // Markdown preview virtualizes sections: move down until the requested embed mounts.
          const preview=app.workspace.activeLeaf.view.containerEl.querySelector('.markdown-preview-view');
          for(let i=0;!el && i<40;i++){
            preview.scrollTop+=500; await new Promise(r=>setTimeout(r,50)); el=document.querySelector(selector);
          }
          if(!el) throw new Error('Embed not mounted: '+EMBED);
          el.scrollIntoView({block:'start'});
          await new Promise(r=>setTimeout(r,350));
          const rows=new Map(),cells=new Map();
          const table=el.querySelector('.bases-table-container');
          const top=table ? preview.scrollTop+table.getBoundingClientRect().top-preview.getBoundingClientRect().top : preview.scrollTop;
          const height=table ? Math.max(table.scrollHeight,table.getBoundingClientRect().height) : 0;
          for(let offset=0;offset<=height;offset+=Math.max(100,preview.clientHeight/2)){
            if(table)table.scrollTop=offset;
            preview.scrollTop=top+offset;
            await new Promise(r=>setTimeout(r,250));
            for(const row of el.querySelectorAll('.bases-tbody .bases-tr')){
              const data=Array.from(row.querySelectorAll('.bases-td')).map(c=>({property:c.dataset.property,text:c.innerText,links:Array.from(c.querySelectorAll('[data-href]')).map(a=>a.dataset.href)}));
              rows.set(JSON.stringify(data[0]),data);
            }
            for(const c of el.querySelectorAll('[data-property="note.text"] .metadata-input-longtext,[data-property="note.summary"] .metadata-input-longtext'))
              cells.set(c.textContent,{clientHeight:c.clientHeight,scrollHeight:c.scrollHeight,whiteSpace:getComputedStyle(c).whiteSpace,overflow:getComputedStyle(c).overflow});
          }
          return JSON.stringify({embed:EMBED,text:el.innerText+'\n'+Array.from(rows.values()).map(r=>r.map(c=>c.text).join('\n')).join('\n'),
            rows:Array.from(rows.values()),cells:Array.from(cells.values()),table:table?{height:table.clientHeight,scrollHeight:table.scrollHeight}:null,
            errors:Array.from(el.querySelectorAll('.bases-error')).map(e=>e.textContent)});
        })()""".replace("EMBED", json.dumps(embed)))

    def set_property(self, path, name, value):
        return self.js("""(async()=>{
          await app.fileManager.processFrontMatter(app.vault.getAbstractFileByPath(PATH),fm=>fm[NAME]=VALUE);
          await new Promise(r=>setTimeout(r,700));
          return JSON.stringify({edited:PATH,property:NAME,value:VALUE});
        })()""".replace("PATH", json.dumps(path)).replace("NAME", json.dumps(name)).replace("VALUE", json.dumps(value)))


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("vault", help="Unique Obsidian vault name")
    parser.add_argument("output", type=Path)
    parser.add_argument("--cli", default=CLI)
    args = parser.parse_args()
    observer = Observer(args.vault, args.cli)
    # Fail closed before any navigation if the chosen vault is not this fixture.
    if not observer.js("JSON.stringify(app.vault.getAbstractFileByPath('bases/prepared-clues.base') !== null)"):
        parser.error("Not a Bases prototype vault")
    evidence = {"version": observer.command("version"), "plugins": observer.command("plugins:enabled", "versions"), "cases": []}
    for path, embeds in [
        ("campaign_1/index/Clues.md", ["bases/clue-index.base#Active", "bases/clue-index.base#Closed"]),
        ("campaign_2/index/Clues.md", ["bases/clue-index.base#Active", "bases/clue-index.base#Closed"]),
        ("world/npcs/Mira.md", ["bases/entity-clues.base"]),
        ("world/npcs/Visitors/Mira.md", ["bases/entity-clues.base"]),
        ("world/locations/Empty.md", ["bases/entity-clues.base"]),
        ("campaign_1/clues/C-1-0001.md", ["bases/clue-sessions.base"]),
        ("campaign_2/clues/C-2-0001.md", ["bases/clue-sessions.base"]),
        ("campaign_1/sessions/S-1-001.md", ["bases/prepared-clues.base", "bases/prepared-locations.base", "bases/prepared-npcs.base"]),
        ("campaign_1/sessions/S-1-003.md", ["bases/prepared-clues.base", "bases/prepared-locations.base", "bases/prepared-npcs.base"]),
    ]:
        opened = observer.open(path)
        evidence["cases"].append({**opened, "views": [observer.capture(e) for e in embeds]})
        args.output.write_text(json.dumps(evidence, indent=2) + "\n")
    args.output.write_text(json.dumps(evidence, indent=2) + "\n")


if __name__ == "__main__":
    main()

File: tests/views/observe.mjs
/** Disposable-vault CDP harness; Node 22+, Obsidian --remote-debugging-port=9225.
 * Uses private UI APIs only for evaluation, never installed in a vault.
 */
import fs from 'node:fs/promises';
const port = process.env.OBSIDIAN_DEBUG_PORT || '9225';
export async function connect(vault) {
  const pages = await (await fetch(`http://127.0.0.1:${port}/json/list`)).json();
  const page = pages.find(p => p.type === 'page' && p.title.includes(vault));
  if (!page) throw Error(`No open review vault: ${vault}`);
  const socket = new WebSocket(page.webSocketDebuggerUrl);
  await new Promise((resolve, reject) => { socket.onopen = resolve; socket.onerror = reject; });
  let id = 0;
  const pending = new Map();
  socket.onmessage = event => {
    const data = JSON.parse(event.data);
    if (pending.has(data.id)) {pending.get(data.id)(data);pending.delete(data.id);}
  };
  async function send(method, params) {
    const current = ++id;
    const answer = new Promise(resolve => pending.set(current, resolve));
    socket.send(JSON.stringify({id: current, method, params}));
    const data = await answer;
    if (data.error) throw Error(JSON.stringify(data.error));
    return data.result;
  }
  async function js(code) {
    const expression = `(async()=>{if(app.vault.getName()!==${JSON.stringify(vault)} || !app.vault.adapter.basePath.startsWith('/private/tmp/armarium25-review/'))throw Error('Not a disposable review vault');return await (${code});})()`;
    const data = await send('Runtime.evaluate', {expression, awaitPromise: true, returnByValue: true});
    if (data.exceptionDetails) throw Error(JSON.stringify(data.exceptionDetails));
    return data.result.value;
  }
  async function open(path) {
    // Foreground rendering must be explicitly requested; default runs never steal focus.
    const focus = process.env.OBSIDIAN_ALLOW_FOCUS === '1'
      ? "const w=require('@electron/remote').getCurrentWindow();w.show();w.focus();" : '';
    return js(`(async()=>{${focus}await app.workspace.getLeaf().openFile(app.vault.getAbstractFileByPath(${JSON.stringify(path)}),{state:{mode:'preview'}});await new Promise(r=>setTimeout(r,650));return document.title;})()`);
  }
  async function capture() {
    return js(`(async()=>{
      const preview=app.workspace.activeLeaf.view.containerEl.querySelector('.markdown-preview-view');
      const views=new Map(),tables=new Map();
      const links=new Set();
      let text='';
      for(let offset=0;offset<=preview.scrollHeight;offset+=Math.max(200,preview.clientHeight/2)) {
        preview.scrollTop=offset;await new Promise(r=>setTimeout(r,80));
        text=preview.innerText;
        for(const el of preview.querySelectorAll('.bases-view')) {
          const src=el.closest('.internal-embed')?.getAttribute('src') || '';
          const rows=[...el.querySelectorAll('.bases-list-item,.bases-tbody .bases-tr')].map(r=>({text:r.innerText,values:[...r.querySelectorAll('input')].map(i=>i.value),links:[...r.querySelectorAll('[data-href]')].map(a=>a.dataset.href)}));
          const key=src+'#'+el.dataset.viewName;
          const previous=views.get(key) || {src,name:el.dataset.viewName,type:el.dataset.viewType,rows:[],errors:[]};
          for(const row of rows)if(!previous.rows.some(r=>r.text===row.text))previous.rows.push(row);
          previous.errors.push(...[...el.querySelectorAll('.bases-error')].map(e=>e.textContent));
          views.set(key,previous);
        }
        for(const el of preview.querySelectorAll('table'))tables.set(el.innerText,{text:el.innerText,links:[...el.querySelectorAll('[data-href]')].map(a=>a.dataset.href)});
        for(const a of preview.querySelectorAll('[data-href]'))links.add(a.dataset.href);
      }
      return {path:app.workspace.getActiveFile().path,title:document.title,text,views:[...views.values()],tables:[...tables.values()],links:[...links],errors:[...preview.querySelectorAll('.bases-error')].map(e=>e.textContent)};
    })()`);
  }
  async function set(path, property, value) {
    return js(`(async()=>{await app.fileManager.processFrontMatter(app.vault.getAbstractFileByPath(${JSON.stringify(path)}),fm=>fm[${JSON.stringify(property)}]=${JSON.stringify(value)});await new Promise(r=>setTimeout(r,1800));return true;})()`);
  }
  async function screenshot(path) {
    const data=await send('Page.captureScreenshot',{format:'png'});
    await fs.writeFile(path,Buffer.from(data.data,'base64'));
  }
  return {js,open,capture,set,screenshot,close:()=>socket.close()};
}

if (process.argv[1] && import.meta.url === new URL(process.argv[1], 'file:').href) {
  const observer = await connect(process.argv[2]);
  try {
    await observer.open(process.argv[3]);
    console.log(JSON.stringify(await observer.capture(),null,2));
  } finally {observer.close();}
}
